File: dart/leet_code/85.dart
/// Dart implementation of the LeetCode problem 85: Maximal Rectangle.
///
/// This module contains a class `Solution` with a method `maximalRectangle`.
/// The `maximalRectangle` method takes in one parameter:
/// - `matrix`: A 2D list of strings representing the binary matrix.
///
/// The method returns an integer representing the area of the maximal rectangle in the binary matrix.
///
/// Error Handling:
/// - This method assumes that the input parameter is well-formed, i.e., `matrix` is a 2D list of strings.
/// - If the input parameter is not well-formed, the behavior of the method is undefined.
library;

import 'dart:math';

class Solution {
  /// Finds the maximal rectangle in a binary matrix.
  ///
  /// This method uses dynamic programming to calculate the maximal width at each point in the matrix.
  /// Then it calculates the maximal area using the calculated widths.
  ///
  /// @param matrix The binary matrix.
  /// @return The area of the maximal rectangle.
  int maximalRectangle(List<List<String>> matrix) {
    // If the matrix is empty, return 0.
    if (matrix.isEmpty) return 0;

    // Get the dimensions of the matrix.
    int m = matrix.length;
    int n = matrix[0].length;

    // Initialize a 2D list to store the maximal width at each point in the matrix.
    List<List<int>> dp = List.generate(m, (_) => List.filled(n, 0));

    // Calculate the maximal width at each point in the matrix.
    for (int i = 0; i < m; i++) {
      for (int j = 0; j < n; j++) {
        if (matrix[i][j] == '1') {
          dp[i][j] = j == 0 ? 1 : dp[i][j - 1] + 1;
        }
      }
    }

    // Initialize the maximal area to 0.
    int maxArea = 0;

    // Calculate the maximal area using the calculated widths.
    for (int i = 0; i < m; i++) {
      for (int j = 0; j < n; j++) {
        if (matrix[i][j] == '1') {
          int width = dp[i][j];
          int area = width;
          for (int k = i - 1; k >= 0; k--) {
            width = min(width, dp[k][j]);
            area = max(area, width * (i - k + 1));
          }
          maxArea = max(maxArea, area);
        }
      }
    }

    // Return the maximal area.
    return maxArea;
  }
}

File: swift/leet_code/85.swift
// Importing Foundation for basic Swift functionalities
import Foundation

/**
 This class provides a solution for finding the maximal rectangle in a binary matrix.

 The `maximalRectangle` function takes a 2D matrix of characters as input, where each character is either '0' or '1'.
 It returns the area of the largest rectangle containing only '1's in the matrix.

 The `largestRectangleArea` function is a helper function that calculates the largest rectangle area in a histogram.
 It takes an array of integers as input, where each integer represents the height of a bar in the histogram.
 It returns the area of the largest rectangle that can be formed in the histogram.
 */
class Solution {
    func maximalRectangle(_ matrix: [[Character]]) -> Int {
        // Initialize the histograms and heightAtColums arrays
        var histograms = Array(repeating: Array(repeating: 0, count: matrix.first?.count ?? 0), count: matrix.count)
        var heightAtColums = [Int: Int]()

        // Iterate over the matrix to calculate the height of '1's at each column
        for row in 0..<matrix.count {
            for column in 0..<matrix[row].count {
                let height = matrix[row][column]
                if height == "1" {
                    heightAtColums[column] = (heightAtColums[column] ?? 0) + 1
                } else {
                    heightAtColums[column] = 0
                }
                histograms[row][column] = heightAtColums[column] ?? 0
            }
        }

        // Initialize the maxArea to the smallest possible integer
        var maxArea = Int.min

        // Iterate over the histograms to calculate the largest rectangle area
        for histogram in histograms {
            let area = largestRectangleArea(histogram)
            maxArea = max(area, maxArea)
        }

        // Return the maxArea
        return maxArea
    }

    func largestRectangleArea(_ heights: [Int]) -> Int {
        // Initialize the stack and maxArea
        var stack = [Int]()
        var maxArea = 0
        var i = 0

        // Iterate over the heights to calculate the largest rectangle area
        while i < heights.count {
            if stack.isEmpty || heights[i] >= heights[stack.last ?? 0] {
                stack.append(i)
                i += 1
            } else {
                let top = stack.removeLast()
                let width = stack.isEmpty ? i : i - stack.last! - 1
                maxArea = max(maxArea, heights[top] * width)
            }
        }

        // Continue to calculate the largest rectangle area with the remaining heights in the stack
        while !stack.isEmpty {
            let top = stack.removeLast()
            let width = stack.isEmpty ? i : i - stack.last! - 1
            maxArea = max(maxArea, heights[top] * width)
        }

        // Return the maxArea
        return maxArea
    }
}
